src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,354 @@
+  def iter_files():
+    for adaptor in all_target_adaptors:
+      if hasattr(adaptor, 'sources'):
+        for file in adaptor.sources.snapshot.files:
+          if file.endswith('.py'):
+            yield file
+
+  return list(iter_files())
+
+
+@dataclass(frozen=True)
+class PytestCoverageReport():
+  report_directory_digest: Digest
+  directory_to_materialize_to: PurePath
+
+
+@rule(name="Generate coverage report")
+async def generate_coverage_report(
+  transitive_targets: TransitiveHydratedTargets,
+  python_setup: PythonSetup,
+  coverage_setup: CoverageSetup,
+  merged_coverage_data: MergedCoverageData,
+  coverage_toolbase: PytestCoverage,
+  source_root_config: SourceRootConfig,
+  subprocess_encoding_environment: SubprocessEncodingEnvironment,
+) -> PytestCoverageReport:
+  """Takes all python test results and generates a single coverage report."""
+  requirements_pex = coverage_setup.requirements_pex
+  # TODO(#4535) We need a better way to do this kind of check that covers synthetic targets and rules extensibility.
+  python_targets = [
+    target for target in transitive_targets.closure
+    if target.adaptor.type_alias == 'python_library' or target.adaptor.type_alias == 'python_tests'
+  ]
+
+  source_roots = source_root_config.get_source_roots()
+  python_files = set(itertools.chain.from_iterable(
+    target.adaptor.sources.snapshot.files for target in python_targets
+  ))
+  coverage_config_content = construct_coverage_config(source_roots, sorted(python_files))
+
+  coveragerc_digest = await Get[Digest](InputFilesContent, get_coveragerc_input(coverage_config_content))
+  sources_digests = [
+    hydrated_target.adaptor.sources.snapshot.directory_digest
+    for hydrated_target in transitive_targets.closure
+    if hasattr(hydrated_target.adaptor, 'sources')
+  ]
+  sources_digest = await Get[Digest](DirectoriesToMerge(directories=tuple(sources_digests)))
+  inits_digest = await Get[InjectedInitDigest](Digest, sources_digest)
+
+
+  merged_input_files: Digest = await Get(
+    Digest,
+    DirectoriesToMerge(directories=(
+      merged_coverage_data.coverage_data,
+      coveragerc_digest,
+      requirements_pex.directory_digest,
+      *sources_digests,
+      inits_digest.directory_digest,
+    )),
+  )
+  coverage_args = [coverage_toolbase.options.report.value]
+  request = requirements_pex.create_execute_request(
+    python_setup=python_setup,
+    subprocess_encoding_environment=subprocess_encoding_environment,
+    pex_path=f'./{coverage_setup.requirements_pex.output_filename}',
+    pex_args=coverage_args,
+    input_files=merged_input_files,
+    output_directories=('htmlcov',),
+    output_files=('coverage.xml',),
+    description=f'Generate coverage report.',
+  )
+
+  result = await Get[ExecuteProcessResult](
+    ExecuteProcessRequest,
+    request
+  )
+  return PytestCoverageReport(result.output_directory_digest, coverage_toolbase.options.report_output_path)
+
+
+class PytestCoverageResult(Goal):
```

Comment:
The goal and its options should be defined in the core, rather than here.

But this is a big divergence from the design doc... this should defintely be a flag on the test goal rather than a separate goal like this.